docs/getting-started/walkthroughs/datastream_api.md

```diff
@@ -0,0 +1,923 @@
+
+        // Check if the flag is set
+        if (lastTransactionWasSmall != null) {
+            if (transaction.getAmount() > LARGE_AMOUNT) {
+                //Output an alert downstream
+                Alert alert = new Alert
+                alert.setId(transaction.getAccountId)
+
+                collector.collect(alert)
+            }
+
+            // Clean up our state
+            flagState.clear()
+        }
+        
+
+
+        if (transaction.getAmount() < SMALL_AMOUNT) {
+            // set the flag to true
+            flagState.update(true)
+        }
+    }
+}
+{% endhighlight %}
+</div>
+</div>
+
+For every transaction, the fraud detector checks the state of the flag for that account.
+Remember, `ValueState` is always scoped to the current key.
+If the flag is non-null, then the last transaction seen for that key was small, and so if the amount for this transaction is large, then the detector outputs a fraud alert.
+
+After that check, the flag state is unconditionally cleared.
+Either the current transaction caused a fraud alert, and the pattern is over, or the current transaction did not cause an alert, and the pattern is broken and needs to be restarted.
+
+Finally, the transaction amount is checked to see if it is small.
+If so, then the flag is set so that it can be checked by the next event.
+
+## Fraud Detector v2: State + Time = &#10084;&#65039;
+
+Many event-driven applications require a strong notion of time to complement their state.
+For example, suppose you wanted to set a 24-hour timeout to your fraud detector; i.e., in the previous example transactions 3 and 4 would only be considered fraud if they occurred within 24 hours of each other.
+Flink allows setting timers which serve as callbacks at some point in time in the future.
+
+Along with the previous requirements:
+
+* Whenever the flag is set to true, also set a timer for 24 hours in the future.
+* When the timer fires, reset the flag by clearing its state.
+* If the flag is ever cleared the timer should be canceled.
+
+To cancel a timer, you have to remember what time it is set for, and remembering implies state, so you will begin by creating a timer state along with your flag state.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+public class FraudDetector extends KeyedProcessFunction<Long, Transaction, Alert> {
+
+	private static final long serialVersionUID = 1L;
+
+	private transient ValueState<Boolean> flagState;
+
+	private transient ValueState<Long> timerState;
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		ValueStateDescriptor<Boolean> flagDescriptor = new ValueStateDescriptor<>(
+				"flag",
+				Types.BOOLEAN);
+
+		flagState = getRuntimeContext().getState(flagDescriptor);
+
+		ValueStateDescriptor<Long> timerDescriptor = new ValueStateDescriptor<>(
+				"timer-state",
+				Types.LONG);
+
+		timerState = getRuntimeContext().getState(timerDescriptor);
+	}
```

Comment:
This whole code uses tabs while the rest is using spaces for indentation. You see this difference also in the HTML output.